rust/src/parse.rs

```diff
@@ -0,0 +1,392 @@
+//! Functions for parsing byte buffers into their MCAP record contents.
+use crate::records::*;
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::error::Error;
+
+/// Returned from [`parse_record`] when parsing fails.
+#[derive(Debug)]
+pub enum ParseError {
+    StringEncoding(std::str::Utf8Error),
+    OpCodeNotImplemented(OpCode),
+    DataTooShort,
+}
+
+impl std::fmt::Display for ParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::DataTooShort => write!(f, "data ended unexpectedly before end of record"),
+            Self::OpCodeNotImplemented(opcode) => write!(f, "opcode {:?} not supported", opcode),
+            Self::StringEncoding(err) => write!(f, "string field not valid utf-8: {}", err),
+        }
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            ParseError::StringEncoding(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+impl From<std::str::Utf8Error> for ParseError {
+    fn from(err: std::str::Utf8Error) -> Self {
+        ParseError::StringEncoding(err)
+    }
+}
+
+fn parse_u16(data: &[u8]) -> Result<(u16, &[u8]), ParseError> {
+    if data.len() < std::mem::size_of::<u16>() {
+        return Err(ParseError::DataTooShort);
+    }
+    let (int_bytes, data) = data.split_at(std::mem::size_of::<u16>());
```

Comment:
is this different from `let data = data[std::mem::size_of::<u16>()..];` (assuming some variant of this is valid...maybe it isn't)?

Reply:
This is just a nice way to bind both slices at once. lines 41-43 are necessary because both split_at and the slice operator will panic on an out-of-bounds index.